fix missing highlighter on OCR recognition results

hwf1324 opened the pull request:
<!-- Please read and fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md
Please also note that the NVDA project has a Citizen and Contributor Code of Conduct which can be found at https://github.com/nvaccess/nvda/blob/master/CODE_OF_CONDUCT.md. NV Access expects that all contributors and other community members read and abide by the rules set out in this document while participating or contributing to this project. This includes creating or commenting on issues and pull requests.

Please initially open PRs as a draft.
When you would like a review, mark the PR as "ready for review".
See https://github.com/nvaccess/nvda/blob/master/.github/CONTRIBUTING.md.
-->

### Link to issue number:
<!-- Use Closes/Fixes/Resolves #xxx to link this PR to the issue it is responding to. -->

### Summary of the issue:

There is no highlighter on the recognition result.

### Description of user facing changes:

There is a highlighter on the recognition result.

### Description of developer facing changes:

Added `api.fakeNVDAObjectClasses` set and `api.isFakeNVDAObject` function to identify fake NVDAObject instances.

### Description of development approach:

Do not use the display model to obtain the caret rectangle of fake NVDAObjects.

### Testing strategy:

Manual testing: After pressing `NVDA+r` to recognize content, move the cursor and observe whether the yellow rectangle follows.

### Known issues with pull request:

By default, the highlight rectangle currently covers words rather than individual characters.

### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

hwf1324 committed:
fix(vision): resolve missing highlighters on recognition results

This change ensures that
`RecogResultNVDAObject` instances are properly ignored when calculating
caret rectangles, allowing highlighters to display correctly.

Fixes an issue where visual enhancements were not applied to content
recognition results.

hwf1324 commented on source/vision/util.py:
```diff
@@ -25,7 +26,11 @@ def getCaretRect(obj: Optional[TextContainerObject] = None) -> locationHelper.Re
+	if (
+		api.isNVDAObject(obj)
+		# Ignore fake NVDAObjects
+		and not isinstance(obj, RecogResultNVDAObject)
```
Should the `isRecogResultNVDAObject` function be created in the `api` module?

CyrilleB79 commented on user_docs/en/changes.md:
```diff
@@ -64,6 +64,7 @@ We recommend using Windows 11, or if that is not possible, the latest Windows 10
 * Battery time announcements now skip redundant "0 hours" and "0 minutes" and use proper singular/plural forms. (#9003, @hdzrvcc0X74)
 * Certain settings will no-longer erroneously be saved to disk when running NVDA from the launcher. (#18171)
 * Incorrect information is no longer displayed in braille when navigating the list of messages in Outlook Classic. (#18993, @nvdaes)
+* Fixed an error where highlighters were not appearing on recognition results. (@hwf1324)
```
The way you wrote it gives the impression that highlighter did not appear on recognition results from time to time.

What about the following instead?
```suggestion
* The browse mode cursor highlighters now appears on recognition results. (@hwf1324)
```

This makes the item more positive.

And rather than a bugfix, I'd rather add this item as a new feature, as to me, this is a new capability.

hwf1324 replied:
I agree.

For me, all English expressions are achieved through translation, so achieving accurate expression is quite difficult.

hwf1324 committed:
Update user_docs/en/changes.md

Co-authored-by: Cyrille Bougot <cyrille.bougot2@laposte.net>

hwf1324 committed:
feat(api): add support for fake NVDAObject detection

- Added `api.fakeNVDAObject` set and `api.isFakeNVDAObject` function to identify fake NVDAObject instances.
- Registered `RecogResultNVDAObject` as a fake NVDAObject in recogUi module.

SaschaCowley commented on source/api.py:
```diff
@@ -500,6 +500,14 @@ def isNVDAObject(obj: Any) -> bool:
 	return isinstance(obj, NVDAObjects.NVDAObject)
 
 
+fakeNVDAObject: set[type[Any]] = set()
```
Why is this necessary?

SaschaCowley replied:
Can this get a more descriptive name, and perhaps a docstring?
```suggestion
fakeNVDAObjectClasses: set[type[Any]] = set()
```

SaschaCowley commented on user_docs/en/changes.md:
```diff
@@ -64,6 +64,7 @@ We recommend using Windows 11, or if that is not possible, the latest Windows 10
 * Battery time announcements now skip redundant "0 hours" and "0 minutes" and use proper singular/plural forms. (#9003, @hdzrvcc0X74)
 * Certain settings will no-longer erroneously be saved to disk when running NVDA from the launcher. (#18171)
 * Incorrect information is no longer displayed in braille when navigating the list of messages in Outlook Classic. (#18993, @nvdaes)
+* The browse mode cursor highlighters now appears on recognition results. (@hwf1324)
```
```suggestion
* The browse mode cursor highlighter now appears on content recognition results, such as when using Windows OCR. (@hwf1324)
```

hwf1324 committed:
Update source/api.py

Co-authored-by: Sascha Cowley <16543535+SaschaCowley@users.noreply.github.com>

hwf1324 committed:
Update user_docs/en/changes.md

Co-authored-by: Sascha Cowley <16543535+SaschaCowley@users.noreply.github.com>

hwf1324 committed:
Update source/api.py

Co-authored-by: Sascha Cowley <16543535+SaschaCowley@users.noreply.github.com>

hwf1324 committed:
refactor(api): improve type hint and documentation for fake NVDA object classes

Updated the type annotation for `fakeNVDAObjectClasses` to be more specific,
indicating it holds types of `documentBase.TextContainerObject`. Also added
documentation explaining that these classes represent virtual NVDAObjects.

hwf1324 committed:
Merge remote-tracking branch 'upstream/master' into pr/hwf1324/19168

hwf1324 committed:
update changes.md

SaschaCowley commented on source/api.py:
```diff
@@ -500,12 +500,14 @@ def isNVDAObject(obj: Any) -> bool:
 
 
-fakeNVDAObject: set[type[Any]] = set()
+fakeNVDAObjectClasses: "set[type[documentBase.TextContainerObject]]" = set()
```
Shouldn't this be
```suggestion
fakeNVDAObjectClasses: set[type[NVDAObjects.NVDAObject]] = set()
```

SaschaCowley commented on source/api.py:
```diff
@@ -500,12 +500,14 @@ def isNVDAObject(obj: Any) -> bool:
-fakeNVDAObject: set[type[Any]] = set()
+fakeNVDAObjectClasses: "set[type[documentBase.TextContainerObject]]" = set()
+"""A collection used to register fake NVDAObject classes.
+These classes are treated as virtual NVDAObjects and may not correspond to actual controls."""
```
```suggestion
"""
A collection used to register fake NVDAObject classes.

These classes are treated as virtual NVDAObjects, and may not correspond to actual controls.
For instance, content recognition results.
"""
```

hwf1324 committed:
Update source/api.py

Co-authored-by: Sascha Cowley <16543535+SaschaCowley@users.noreply.github.com>

hwf1324 committed:
Update source/api.py

Co-authored-by: Sascha Cowley <16543535+SaschaCowley@users.noreply.github.com>

hwf1324 committed:
Update user_docs/en/changes.md

Co-authored-by: Sascha Cowley <16543535+SaschaCowley@users.noreply.github.com>

hwf1324 committed:
Update user_docs/en/changes.md

Co-authored-by: Sascha Cowley <16543535+SaschaCowley@users.noreply.github.com>

SaschaCowley reviewed (approved):
Thanks, @hwf1324

SaschaCowley committed:
Merge remote-tracking branch 'origin/master' into RecogResultHighlighter

# Conflicts:
#	user_docs/en/changes.md
